helix-rest/src/main/java/org/apache/helix/rest/server/resources/metadatastore/MetadataStoreDirectoryAccessor.java

```diff
@@ -90,9 +90,13 @@ public Response getAllMetadataStoreRealms() {
   @Path("/metadata-store-realms/{realm}")
   public Response addMetadataStoreRealm(@PathParam("realm") String realm) {
     try {
-      _metadataStoreDirectory.addMetadataStoreRealm(_namespace, realm);
+      if (!_metadataStoreDirectory.addMetadataStoreRealm(_namespace, realm)) {
+        return serverError();
+      }
     } catch (IllegalArgumentException ex) {
       return notFound(ex.getMessage());
+    } catch (IllegalStateException ex) {
+      return serverError();
```

Comment:
Where are you throwing an IllegalStateException?

Reply:
In `ZkRoutingDataWriter` if `_zkClient.isClosed()`.

Reply:
If that’s the case, I would put it in addMetadatStoreRealm or a proper place, but not here in this endpoint logic. It is not endpoint logic’s responsibility to process such exception.

Reply:
+1. Let's make sure the MSDS server logs the errors. it would be helpful if we could include as much detail as possible including the host name, etc.

Reply:
If it's not handled, the exception will cause a server error anyways. Are you suggesting to handle it and not raise server error? IMO zkClient being closed is definitely a server error.

Reply:
You could let this IllegalStateException float up all the way to the servlet..

For instance, for uncaught exceptions, the servlet container usually catch the Exception, logs it and returns http status 500.

Just make sure MetadataStoreDirectory logs the exception properly.

Reply:
I see. I'll remove this section of the code. Thank you both for the insight.